datalad_next/tree.py

```diff
@@ -0,0 +1,1130 @@
+            If paired with [CMD: --dataset-depth CMD][PY: dataset_depth PY],
+            refers to the maximum directory level to display underneath each
+            dataset.""",
+            constraints=EnsureInt() & EnsureRange(min=0) | EnsureNone()),
+        dataset_depth=Parameter(
+            args=("--dataset-depth",),
+            doc="""maximum level of nested subdatasets to display""",
+            constraints=EnsureInt() & EnsureRange(min=0) | EnsureNone()),
+        include_files=Parameter(
+            args=("--include-files",),
+            doc="""include files in output display""",
+            action='store_true'),
+        include_hidden=Parameter(
+            args=("--include-hidden",),
+            doc="""include hidden files/directories in output display""",
+            action='store_true'),
+    )
+
+    _examples_ = [
+        dict(text="Display up to 3 levels of the current directory's "
+                  "subdirectories and their contents",
+             code_py="tree(depth=3, include_files=True)",
+             code_cmd="datalad tree --depth 3 --include-files"),
+        dict(text="Display all first- and second-level subdatasets of "
+                  "datasets located anywhere under /tmp (including in hidden "
+                  "directories) regardless of directory depth",
+             code_py="tree('/tmp', dataset_depth=2, include_hidden=True)",
+             code_cmd="datalad tree /tmp --dataset-depth 2 --include-hidden"),
+        dict(text="Display first- and second-level subdatasets and their "
+                  "contents, up to 1 directory deep within each dataset",
+             code_py="tree(dataset_depth=2, depth=1)",
+             code_cmd="datalad tree --dataset-depth 2 --depth 1"),
+    ]
+
+    @staticmethod
+    @datasetmethod(name='tree')
+    @eval_results
+    def __call__(
+            path='.',
+            *,
+            depth=None,
+            dataset_depth=None,
+            include_files=False,
+            include_hidden=False):
+
+        if dataset_depth is not None:
+            # special tree defined by subdataset nesting depth
+            tree_cls = DatasetTree
+            dataset_tree_args = {"max_dataset_depth": dataset_depth}
+        else:
+            # simple tree defined by directory depth
+            tree_cls = Tree
+            dataset_tree_args = {}
+
+        tree = tree_cls(
+            Path(path),
+            max_depth=depth,
+            exclude_node_func=build_excluded_node_func(
+                include_hidden=include_hidden, include_files=include_files),
+            **dataset_tree_args
+        )
+
+        for node in tree.generate_nodes():
+            # yield one node at a time to improve UX / perceived speed
+            res_dict = {
+                "action": "tree",
+                "path": str(node.path),
+                "type": node.TYPE,
+                "depth": node.depth,
+                "exhausted_levels": list(tree.exhausted_levels),
```

Comment:
Here is an example result record from a deep tree, but for toplevel-ish directory:

```json
{
  "action": "tree",
  "count": {
    "datasets": 5,
    "directories": 9469
  },
  "dataset_abs_depth": 0,
  "dataset_depth": 0,
  "dataset_is_installed": true,
  "depth": 1,
  "exhausted_levels": [
    2,
    7,
    9,
    10,
    11,
    12,
    13,
    8,
    5,
    3,
    6,
    4
  ],
  "path": "/home/mih/dcm",
  "status": "ok",
  "type": "dataset"
}
```

If I get it right, the `exhausted_levels` are reported to understand for which branches of the tree to no longer draw `| ` prefix markers.

But in the case of a `depth=1` record, it should not be necessary to report any `exhausted_levels`, because they should all be exausted. In general, any report that is larger than the present `depth` is unnecessary.

As a "cosmetic" fix, we could do something like

```suggestion
                "exhausted_levels": [l for l in tree.exhausted_levels if l < node.depth]
```

but I am wondering whether this set should not be pruned in the tree node generator itself?

Reply:
Yes, that bothered me too! `exhausted_levels` carried around the entire history of the Tree. It should be scoped to the current node instead.

Fixed it in the node generator:
https://github.com/datalad/datalad-next/blob/6db0e12fdef117c53f4dd7c84fa455e7e740596a/datalad_next/tree.py#L728-L733

Added test:
https://github.com/datalad/datalad-next/blob/6db0e12fdef117c53f4dd7c84fa455e7e740596a/datalad_next/tests/test_tree.py#L528-L531

Reply:
Nice! Thx!